cmd/src/dump.rs

```diff
@@ -0,0 +1,101 @@
+use engine_rocks::raw_util::CFOptions;
+use engine_rocks::{
+    self,
+    raw::{DBOptions, Env},
+    RocksEngine,
+};
+use engine_traits::{Iterable, RaftEngine, CF_DEFAULT, CF_RAFT};
+use kvproto::raft_serverpb::RaftLocalState;
+use protobuf::Message;
+use raft::eraftpb::Entry;
+use std::collections::HashMap;
+use std::path::PathBuf;
+use std::sync::Arc;
+
+pub type RaftDumpData = HashMap<u64, (Vec<Entry>, RaftLocalState)>;
+
+/// Check the potential original raftdb directory and try to dump data out.
+///
+/// TiKV only looks for the name in the directory where `raft-engine` is located,
+/// and assumes that `raft` is original raftdb directory.
+pub fn check_and_dump_raft_db(new_path: &str, env: Arc<Env>) -> Option<RaftDumpData> {
+    let origin_path = get_raftdb_path(new_path);
+    if !RocksEngine::exists(&origin_path) {
+        return None;
+    }
+    let mut opt = DBOptions::default();
+    opt.set_env(env);
+    let ct_opt = vec![CFOptions::new(CF_DEFAULT, Default::default())];
+    let db = engine_rocks::raw_util::new_engine_opt(&origin_path, opt, ct_opt)
+        .unwrap_or_else(|s| fatal!("failed to create origin raft engine: {}", s));
+    let origin_engine = RocksEngine::from_db(Arc::new(db));
+    let dump_data = dump_data_from_raft_rocksdb(&origin_engine);
+    if dump_data.is_empty() {
+        return None;
+    }
+    Some(dump_data)
+}
+
+/// Recover original raft data into new raft-engine.
+pub fn dump_data_into_raft_engine<E: RaftEngine>(
+    raft_db: &E,
+    new_path: &str,
+    data: Option<RaftDumpData>,
+) {
+    if let Some(data) = data {
+        for (id, (entries, state)) in data {
+            raft_db.put_raft_state(id, &state).unwrap();
```

Comment:
It's better to use a log batch for every region. And then sync for every log batch.